### PR TITLE
ci: run benchmarks on every pr and alert if perf is worse

### PR DESCRIPTION
**Pull request by miparnisari**
Close https://github.com/mailgun/gubernator/issues/199

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -18,6 +18,10 @@ test:
 		go tool cover -html coverage.out -o coverage.html; \
 		exit $$ret)
 
+.PHONY: bench
+bench:
+	go test ./... -bench . -benchtime 5s -timeout 0 -run=XXX -cpu 1 -benchmem
```
Comment (thrawn01) [inline Makefile:23]: Some of the tests bench concurrent access. I don't think we need `-run=XXX -cpu 1`

**Review by thrawn01 — COMMENTED** [review]
Thank you for taking the time to make this PR! This is a great addition. I'm not sure yet why the bench test timed out at 30 minutes. But we should probably timeout the test after 10 minutes.

File: Makefile (hunk 1 above)
Comment (miparnisari) [reply Makefile]: Ahh OK. I'll remove `-cpu 1` but keep `-run` since that is supposed to ensure that all benchmarks in all folders run

**Comment by miparnisari** [comment]
> I'm not sure yet why the bench test timed out at 30 minutes. 

@thrawn01 I think either `BenchmarkLRUCache/Concurrent_writes` or https://github.com/miparnisari/gubernator/blob/61e6cd7bd0e2d876431b8d085c5d0226ea16a894/lrucache_test.go#L551 timed out. Probably because it executed all the reads serially? 🤷‍♀️ 

> But we should probably timeout the test after 10 minutes.

Each benchmark is configured to run for 5 seconds and the benchmark step in the github action are set to either 15 or 30 minutes

File: Makefile (hunk 1 above)
Comment (Baliedge) [reply Makefile]: nit: I've used `-run '^$'` for this purpose to match an empty string which is (probably) impossible.

File: .github/workflows/on-pull-request.yml [hunk 2]
```diff
@@ -44,3 +44,55 @@ jobs:
...
+        uses: actions/cache/restore@v3
+        with:
+          path: ./cache/benchmark-data.json
+          key: ${{ steps.get-main-branch-sha.outputs.sha }}-${{ runner.os }}-go-benchmark
+          fail-on-cache-miss: true
+
+      - name: Compare benchmarks
+        uses: benchmark-action/github-action-benchmark@v1
+        if: success()
+        with:
```
Comment (Baliedge) [inline .github/workflows/on-pull-request.yml:84]: Please add `alert-threshold` setting here so it's clear what the performance threshold is.

**Review by Baliedge — CHANGES_REQUESTED** [review]

File: .github/workflows/on-pull-request.yml [hunk 3]
```diff
@@ -44,3 +44,55 @@ jobs:
...
+
+      - name: Run benchmark and store the output to a file
+        run: |
+          set -o pipefail
+          make bench | tee ${{ github.sha }}_bench_output.txt
+
+      - name: Get Main branch SHA
+        id: get-main-branch-sha
+        run: |
+          SHA=$(git rev-parse origin/main)
```
Comment (Baliedge) [inline .github/workflows/on-pull-request.yml:71]: We use `master` branch, not `main`.

**Review by Baliedge — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (miparnisari) [reply Makefile]: I don't know what i'm doing wrong, but i get an error

```
[21/11/23 4:36:35] ~/GitHub/gubernator (run-benchmarks) $ make bench
go test ./... -bench . -benchtime 5s -timeout 0 -run '^ -benchmem
/bin/sh: -c: line 0: unexpected EOF while looking for matching `''
/bin/sh: -c: line 1: syntax error: unexpected end of file
make: *** [bench] Error 2
[21/11/23 4:36:44] ~/GitHub/gubernator (run-benchmarks) $ 
```

**Comment by Baliedge** [comment]
@miparnisari Could we test the `go-bench` check to show it fails when exceeding the threshold?  Maybe make a temporary change to add sleeps in the benchmarks?

**Comment by miparnisari** [comment]
@Baliedge i would need to merge this first into `master` so that I have something to compare with :) But i can do that in a different PR.

**Review by thrawn01 — APPROVED** [review]
Looks good to me, Does this require a GitHub API key in secrets?

**Comment by thrawn01** [comment]
Never mind, I just answered my own question 😜

**Pull request merged**